sklearn/preprocessing/_encoders.py

```diff
@@ -764,10 +776,51 @@ def fit(self, X, y=None):
         self
 
         """
-        # base classes uses _categories to deal with deprecations in
-        # OneHoteEncoder: can be removed once deprecations are removed
-        self._categories = self.categories
-        self._fit(X)
+        X_temp = check_array(X, force_all_finite=False, dtype=None)
+        if not hasattr(X, 'dtype') and np.issubdtype(X_temp.dtype, np.str_):
+            X = check_array(X, force_all_finite=False,
+                            dtype=np.object)
+        else:
+            X = X_temp
+
+        n_samples, n_features = X.shape
+
+        if self.categories != 'auto':
+            if X.dtype != object:
+                for cats in self.categories:
+                    if not np.all(np.sort(cats) == np.array(cats)):
+                        raise ValueError("Unsorted categories are not "
+                                         "supported for numerical categories")
+            if len(self.categories) != n_features:
+                raise ValueError("Shape mismatch: if n_values is an array,"
+                                 " it has to be of shape (n_features,).")
+
+        self.categories_ = []
+
+        for i in range(n_features):
+            Xi = X[:, i]
+            if self.categories == 'auto':
+                try:
+                    cats = _encode(Xi)
+                except TypeError:
```

Comment:
What case does this handle? It looks very suspicious, but if we're to do it, it at least needs a comment.